### ISSUE TITLE
type checking CI is failing

### ISSUE THREAD

**Issue by keewis** [pydata/xarray#7270]
The most recent runs of the type checking CI have started to fail with a segfault:
```
/home/runner/work/_temp/dac0c060-b19a-435a-8063-bbc5b8ffbf24.sh: line 1:  2945 Segmentation fault      (core dumped) python -m mypy --install-types --non-interactive --cobertura-xml-report mypy_report
```
This seems to be due to the release of `mypy=0.990`.

#7269 pinned `mypy` to `mypy<0.990`, which should be undone once we figured out how to fix that (most likely by waiting on the next release), and any complaints the new version has (see this [workflow run](https://github.com/pydata/xarray/actions/runs/3412877230/jobs/5678944346)).

**Comment by headtr1ck** [comment]
Locally (Windows) it works for me.

Only lots of new errors, most importantly the change of mypy to `no_implicit_optional=True`.
Guess we can just fix this in a PR, I like the explicit defaults anyway better.

**Comment by keewis** [comment]
should this have been closed by #7283?

**Comment by headtr1ck** [comment]
> should this have been closed by #7283?

Too long ago...
I assume I had a reason why I didn't close this issue?
